Show/edit run policy for a BuildConfig

jhadvig opened the pull request:
Based on https://trello.com/c/qVzqRo0t/654
Things we want:
- show/edit buildConfigs `runPolicy` 
- get rid of disabling `Start Build` and `Rebuild` buttons (based on https://github.com/openshift/origin/issues/8510)

Things we might want:
- bulk actions on BCs - start-build
- bulk actions on Builds - rebuild, cancel-build

Preview of showing BS's runPolicy:
![screenshot-34](https://cloud.githubusercontent.com/assets/1668218/14984206/d6cf281a-1141-11e6-98a5-65ddde535aec.png)
![screenshot-35](https://cloud.githubusercontent.com/assets/1668218/14984196/c6190d4c-1141-11e6-9898-0d0eea2af252.png)


And here are couple of drafts for the BC's edit page. In here I've created a separate section, since it wouldnt make sense to have the `runPolicy` in any of the other section(source config., image config.,) :
1. dropdown with a description under
![screenshot-31](https://cloud.githubusercontent.com/assets/1668218/14984261/0c8fd9f4-1142-11e6-9536-e4695b31d07b.png)
2. dropdown with a italic description under
![screenshot-30](https://cloud.githubusercontent.com/assets/1668218/14984286/2ea537d2-1142-11e6-9339-78d6b70ef483.png)
3. radio buttons with tooltip for description
![screenshot-33](https://cloud.githubusercontent.com/assets/1668218/14984297/3b28a0b6-1142-11e6-8342-290ace000efe.png)


Drafts 1 and 2 contain tooltip next to the section header that says `The build run policy describes the order in which the builds created from the build configuration should run`. Draft 3 wont have the scetion header tooltip just a tooltip to each radio button, explaining the given runPolicy type.

Also I've haven't done any effort to show the `runPolicy` type for the build show page, since the type is a part of the build labels and those we are already listing, so I think thats enough.

@jwforres @spadgett PTAL. We also need to align on whether we want to have the bulk actions I've described at the top. I think they could come really handy, mainly the bulk cancellation of builds.
Thoughts ?

jhadvig commented:
cc'ing @mfojtik since `runPolicy` is his baby and he might have some valuable input on this

mfojtik commented:
@jhadvig looks pretty good 👍

jwforres commented:
i would probably put run policy underneath triggers, to me it is less important than env vars and triggers because its not something someone will commonly change

jwforres commented:
i like the dropdown with the regular help text underneath (not italics)

jwforres commented:
i think everyone agrees in theory that we want bulk actions, but i dont think it should be part of this PR

spadgett commented:
> i like the dropdown with the regular help text underneath (not italics)

+1

jhadvig commented:
@jwforres I've addressed your comments.
But I've found a small issue that I would like to ask you for a proper fix. The thing is that then you click on the `Start build` button on BC page or `Rebuild` button on the Build page it stays focused.
So what I'm trying to do is to call the `.blur()` method on the button element when its clicked (see the second commit) ... but I'm not sure its the right approach so wanted to ask you..
Thanks so much ! :)

jhadvig commented:
@jwforres I've opend an issue related to the problem with `focus` https://github.com/openshift/origin/issues/8756

jwforres commented on assets/app/scripts/controllers/build.js:
```diff
@@ -48,14 +48,6 @@ angular.module('openshiftConsole')
       $scope.logCanRun = !(_.includes(['New', 'Pending', 'Error'], build.status.phase));
     };
 
-    var updateCanBuild = function() {
```
i think there is still a case where the a button should be disabled.  If you are looking at a build whose build config has been deleted, will Rebuild work?

jhadvig replied:
yep you are right, it will return error

jwforres commented on assets/app/scripts/controllers/build.js:
```diff
@@ -153,7 +143,8 @@ angular.module('openshiftConsole')
         $scope.cloneBuild = function() {
           var name = _.get($scope, 'build.metadata.name');
-          if (name && $scope.canBuild) {
+          $('.clone-build-btn').blur();
```
go ahead and take this out, if we decide to do this we will do it everywhere at once

jwforres commented on assets/app/scripts/controllers/buildConfig.js:
```diff
@@ -165,17 +155,8 @@ angular.module('openshiftConsole')
-
-        $scope.cloneBuild = function(buildName) {
-          BuildsService.cloneBuild(buildName, context, $scope);
+          $('.start-build-btn').blur();
```
take out

jwforres commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -107,6 +107,23 @@ angular.module('openshiftConsole')
       builderImageChangeTrigger: {},
       imageChangeTriggers: []
     };
+    $scope.runPolicyTypes = [
```
we typically don't put something like this in the controllers, we use an ng-switch on the type in the view

jwforres commented on assets/app/views/browse/build-config.html:
```diff
@@ -300,6 +297,19 @@
+                                  {{buildConfig.spec.runPolicy | sentenceCase}}
+                                  <span class="help action-inline">
+                                    <a href>
+                                      <i ng-if="buildConfig.spec.runPolicy === 'Serial'" class="pficon pficon-help" data-toggle="tooltip" aria-hidden="true" data-original-title="Builds triggered from this Build Configuration will run one at the time, in the order they have been triggered."></i>
```
dont use ng-if's here, use an ng-switch

jhadvig commented:
@jwforres I've update the PR
Had to move back the watcher in the buildConfig.js controller, since it was preventing from loadingthe list of builds even if the BC was gone. Spoke with @benjaminapetersen about it.

jwforres commented on assets/app/scripts/controllers/buildConfig.js:
```diff
@@ -149,8 +82,43 @@ angular.module('openshiftConsole')
 
+      watches.push(DataService.watch("builds", context, function(builds, action, build) {
+        $scope.emptyMessage = "No builds to show";
+        // TODO we should send the ?labelSelector=buildconfig=<name> on the API request
```
can you actually take this comment out, it should have been removed as part of @benjaminapetersen PR that added the selector but i missed it

benjaminapetersen replied:
Oop, my bad, missed it too.

jwforres commented on assets/app/scripts/controllers/buildConfig.js:
```diff
@@ -175,17 +143,7 @@ angular.module('openshiftConsole')
-          }
-        };
-
-        $scope.cancelBuild = function(build, buildConfigName) {
```
assuming these werent actually being used

jhadvig replied:
no they weren't

jwforres commented on assets/app/views/browse/build-config.html:
```diff
@@ -300,6 +297,19 @@
+                                <dd>
+                                  {{buildConfig.spec.runPolicy | sentenceCase}}
+                                  <span class="help action-inline">
+                                    <a href ng-switch="buildConfig.spec.runPolicy">
```
if they add a new runPolicy that we don't know about we could end up with an empty `<a>` here, I would add an ng-switch-default that says `Builds triggered from this Build Configuration will run using the {{buildConfig.spec.runPolicy | sentenceCase}} policy.` its not the most helpful help in that case but it would at least not be broken :)

jhadvig replied:
Good catch :+1:

jwforres commented on assets/app/views/edit/build-config.html:
```diff
@@ -578,6 +578,27 @@
+                            ng-options="type as (type | sentenceCase) for type in runPolicyTypes">
+                          </select>
+                        </div>
+                        <div ng-switch="updatedBuildConfig.spec.runPolicy">
```
not as concerned about an ng-switch-default here since any empty div isnt really a problem, but you could do something similar here

jhadvig replied:
will do :)

jwforres commented on assets/app/views/edit/build-config.html:
```diff
@@ -578,6 +578,28 @@
+                          <div ng-switch-when="Serial">Builds triggered from this Build Configuration will run one at the time, in the order they have been triggered.</div>
+                          <div ng-switch-when="Parallel">Builds triggered from this Build Configuration will run all at the same time. The order in which they will finish is not guranteed.</div>
+                          <div ng-switch-when="SerialLatestOnly">Builds triggered from this Build Configuration will run one at the time. When a currently running build completes, the next build that will run is the latest build created. Other queued builds will be cancelled.</div>
+                          <div ng-switch-default>Builds triggered from this Build Configuration will run using the {{buildConfig.spec.runPolicy | sentenceCase}} policy.</div>
```
updatedBuildConfig not buildConfig

jwforres commented:
[merge]
